src/stdlib/base.rs

```diff
@@ -8,6 +8,76 @@ use crate::{
 };
 
 pub fn load_base<'gc>(ctx: Context<'gc>) {
+    ctx.set_global(
+        "tonumber",
+        Callback::from_fn(&ctx, |ctx, _, mut stack| {
+            fn extract_number_data(bytes: &[u8]) -> (&[u8], i64) {
+                let first_nonspace = bytes.iter().position(|b| !b.is_ascii_whitespace());
+                let last_nonspace = bytes.iter().rposition(|b| !b.is_ascii_whitespace());
+                let bytes = match (first_nonspace, last_nonspace) {
+                    (Some(f), Some(l)) => &bytes[f..=l],
+                    (Some(f), None) => &bytes[f..],
+                    (None, Some(l)) => &bytes[..=l],
+                    (None, None) => bytes,
+                };
+
+                let is_neg = bytes.first().is_some_and(|b| *b == b'-');
+                let bytes = if is_neg || bytes.first().is_some_and(|b| *b == b'+') {
+                    &bytes[1..]
+                } else {
+                    bytes
+                };
+                let sign = if is_neg { -1 } else { 1 };
+                (bytes, sign)
+            }
+
+            if stack.is_empty() {
+                Err("Missing argument(s) to tonumber".into_value(ctx))?
+            } else if stack.len() == 1 {
+                let prenumber = stack.consume::<Value>(ctx)?;
+                stack.replace(ctx, prenumber.to_numeric().unwrap_or(Value::Nil));
+            } else {
+                let (s, base) = stack.consume::<(String, i64)>(ctx)?;
+                let (bytes, sign) = extract_number_data(s.as_bytes());
+                stack.replace(ctx, {
+                    if !(2..=36).contains(&base) {
+                        Err("base out of range".into_value(ctx))?;
+                    }
+                    // Since a base [2 - 36 are the only valid values] was specified, this *must*
+                    // represent an integer of the base
+                    const BASE_ELEMENTS: &[u8; 36] = b"0123456789abcdefghijklmnopqrstuvwxyz";
+                    let byte_values = bytes
+                        .iter()
+                        .map(|b| {
+                            BASE_ELEMENTS
+                                .iter()
+                                .position(|i| b.to_ascii_lowercase() == *i)
+                        })
+                        .collect::<Vec<_>>();
+                    if byte_values
+                        .iter()
+                        .any(|v| v.is_none() || v.is_some_and(|v| v >= base as usize))
+                    {
+                        Value::Nil
+                    } else {
+                        let mut value: i64 = 0;
+                        for (idx, elem) in byte_values.into_iter().rev().enumerate() {
+                            let Some(elem) = elem else {
+                                // The earlier byte value check means that this should not be hit
+                                unreachable!();
```

Comment:
it's probably fine if you just want to `return Ok(Value::Nil)` here instead of doing a pre-check pass, which might be a lot simpler.